### ISSUE TITLE
Checklist

### ISSUE THREAD

**Issue by radiantly** [C0derLint/js-org-ci#5]
## CI Checklist  
 - [x] **Handle all PRs**
   - [x] **Adding** a record
   - [x] **Modifying** an existing record
   - [x] **Removing** a record
 - [x] **Title is whatever.js.org**
 - [x] **PR only modifies cnames_active.js**
 - [x] **PR correctly modifies cnames_active.js**
 - [x] **Any new/modified records are in alphabetical order**
 - [x] **No duplicate records**
   - [x] **No duplicate `*.js.org` subdomain names**
   - [x] **Different subdomain names should not point to a same site** Should we check for this?
 - [x] **cnames_active.js retains validity for JS**
 - [x] **no usage of // noCF? (don´t add this in a new PR) on new/modified entry lines**
 - [x] **check for CNAME** (redirect check) if target requested is a github.io domain on new/modified entry
 - [x] **Disallow restricted names** Looking at the cnames_restricted.js file, it looks intended to be read by a person over code ..however we could use regex to read it.
 - [x] **Automatically add labels to PR.** This and the following 2 features can only be added if the bot is given write permissions. This possibly may not be safe as the API key could be extracted from the (Travis) CI environment. On Circle CI however, we could simply hide build logs (I think).
 - [ ] **Automatically make required changes** If the PR title is wrong, but everything else is right, modify it for the user. Same goes, for silly mistakes like trailing slashes, and such.
 - [x] **Request changes** from the person who opened the PR.
 - [ ] **Take and append a screenshot to the end site PR comment.** This will decrease the speed of the test. Also, I'm not sure if we can actually do this because of the redirects
 - [x] **Move to Circle CI** So that we can actually safeguard the API key unlike Travis CI

**Comment by radiantly** [comment]
TODO: Investigate what this does :laughing: 
```js
// check for alphabetical order
    console.info("TEST: alphabetical order");
    linesNew.forEach((line, i) => {
      if (i)
        if(!(`${line}`.localeCompare(`${linesNew[i - 1]}`) !== -1))
          throw `You should keep the list in alphabetical order`;
        
});
```
**EDIT:** Figured it out, it was that `if(i)` that threw me off lol

**Comment by MattIPv4** [comment]
@C0derLint That is basically going through each line in the js and comparing it to the previous line to ensure it is still alphabetically ordered.

**Comment by radiantly** [comment]
@MattIPv4 I've added the alphabetic sort snippet, fixed the PR title regex, and changed the wording of the noCF condition.

**Comment by radiantly** [comment]
@MattIPv4 Would you like me to add anything else before you check the code?

**Comment by MattIPv4** [comment]
I've left some feedback on some of your test PRs here.
In addition:

`Automatically add labels to PR. This feature can only be added if the bot is given write permissions, and is not safe since the API key could be extracted from the CI environment,`
What are the chances of the key getting extracted, how would this even be done?
If this is a serious security issue, could it just list which labels need to be added by a human?

**Comment by radiantly** [comment]
@MattIPv4 Thank you for your feedback. About the API key: Travis CI doesn't allow any secure way of keeping it safe, while I believe on Circle CI, we could simply disable build logs or something similar. https://github.com/facebook/react-native uses danger-bot and Circle CI and doesn't have any problems, so we'll probably have to migrate to Circle CI. Will have to do a bit more research though.

**Comment by radiantly** [comment]
@MattIPv4 I've moved to Circle CI, and it's pretty great .. after tweaking a few settings, it seems like the API key is safe in the CI. From your previous comments, it can now handle modification and removal of records and also parses the restricted cnames file. However, what about the repo maintainer? @indus

**Comment by MattIPv4** [comment]
@C0derLint I'll have play looking through the code, your PRs and submitting some of my own in a bit. Looks good though. Are labels working automatically and actually submitting a review instead of a comment?

**Comment by radiantly** [comment]
@MattIPv4 Bot now adds labels to PRs. Reading the [docs](https://octokit.github.io/rest.js/#api-Pulls) to find out how I can request changes.

**EDIT:** Seems way more complicated than needed. First a review needs to be created with `createReview()` with the appropriate line number, then submitted using `submitReview()` with the `REQUEST_CHANGES` flag. Moreover, this opens up quite a few edge cases. We could add this in the future though.

**Comment by MattIPv4** [comment]
@C0derLint You should just be able to make a general review instead of a line specific review?

Edit: You use any of the three flags as appropriate, if the PR is good to go use the `APPROVE` flag and if there are issues use the `REQUEST_CHANGES`. If the bot reaches a situation where it doesn't know what to do, it can simply use `COMMENT` review instead.

**Comment by radiantly** [comment]
@MattIPv4 I'm looking at the docs once again, I can't seem to understand where I can retrieve review_id for https://octokit.github.io/rest.js/#api-Pulls-submitReview

**EDIT:** From what I understand, I need to `createReview()`, `listReviews()` and then `submitReview()` which seems quite convoluted to do this simple thing.

**EDIT2:** Figured it out :D

**EDIT 3:** Backtrack on that, I've got no idea how to implement this straight from danger, I'm going to open an issue at danger-js asking for their guidance. 

**EDIT 4:** Let's see what they say. Worst case scenario we could use the GitHub API natively.

**Comment by indus** [comment]
> However, what about the repo maintainer? @indus

@C0derLint First of all -  Thank you for your afford with this. It sounds like you are trying to bring our current implementation based on Travis CI to the next level.

But I'm not sure your point about Travis Ci and secure API keys is correct.
> About the API key: Travis CI doesn't allow any secure way of keeping it safe

This sounds like it is possible: https://docs.travis-ci.com/user/environment-variables/#defining-encrypted-variables-in-travisyml

While I can imagine to set this up with any other CI solution (as long as it is secure and it has a free plan for projects like JS.ORG) I have to say I would feel more confident if we could stick to Travis CI (just for its great reputation and its superb Github integration).

I also would have liked to first gather some experience with just a *passiv* implementation (just all the checks with warnings only and no modifications like "Automatically make required changes" in your checklist). I would guess this should also be possible without API keys.

**Comment by radiantly** [comment]
@indus Thank you for your reply. 

> This sounds like it is possible: https://docs.travis-ci.com/user/environment-variables/#defining-encrypted-variables-in-travisyml

This cannot work for us since these "encrypted" variables are not decrypted in PR builds (from forks) -- this is because anyone who submits a PR could add something like `echo $SECRET_VAR` in the .travis.yml

I understand that sticking to Travis CI is desirable, but in this case, using Circle CI's free plan suits our needs better as we can disable build logs for unauthorized users and hide it more effectively. It's almost as popular - [source](https://github.blog/2017-11-07-github-welcomes-all-ci-tools/).

> I also would have liked to first gather some experience with just a passive implementation (just all the checks with warnings only and no modifications like "Automatically make required changes" in your checklist). I would guess this should also be possible without API keys.

I haven't even started working on this feature, so no problems. However, an API key still needs to be used so that the machine/bot account can comment. See an example PR #25

**Comment by MattIPv4** [comment]
> Different subdomain names should not point to a same site Should we check for this?

Yes, I believe we should? @indus thoughts?

> Request changes from the person who opened the PR.

Any progress on this @C0derBot ?

**Comment by radiantly** [comment]
> Yes, I believe we should? @indus thoughts?

I've added this feature anyway in f7fe80b

> > Request changes from the person who opened the PR.

> Any progress on this @C0derLint?

Well, no. might have to do this straight using the Github API.

**Comment by radiantly** [comment]
@MattIPv4 Done.

**Comment by MattIPv4** [comment]
@C0derLint Would it be possible to make the results themselves the approval/request changes message?

Edit: For clarification, it would be best if when it requested changes, it simply listed the fails and warnings that need to be resolved. Once it then approves changes, it can list all the tests passed.

**Comment by radiantly** [comment]
@MattIPv4 No, I don't think so. ATM, we use [danger/danger-js](https://github.com/danger/danger-js) and it doesn't really support this feature.

Maybe instead, the Request changes comment could explain the error, like you do:
> Hi there!
> Before we can accept your domain request, you need to create a file named CNAME in the root of your github pages site (this may not be the root of your repository). memo
This file should contain a single line with the full js.org domain you are requesting, in this instance mydomain.js.org.

This would help users that don't really understand the error generated.

**Comment by MattIPv4** [comment]
@C0derLint Hm okay, would it be possible to stop danger posting a comment at all then and just internally using the warnings etc that it raises for the custom request changes etc?

**Comment by radiantly** [comment]
@MattIPv4 Technically, yes. With the `-t` flag, we could prevent it from posting on the PR. Then, we could list the required changes on the review.

Busy this weekend though, I'm not sure if I can work on this.

**Comment by radiantly** [comment]
@MattIPv4 Finally got back, changed a bit of code, and #35, #36.

**Comment by MattIPv4** [comment]
Hiya @radiantly - I'll be honest, I completely forgot about this until today when cleaning up some old issues on my account.

What's the situation with it now, any progress on the automatic changes?
